<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>netReg commandline tutorial • netReg</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/lumen/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">netReg</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/netReg_R.html">R</a>
</li>
<li>
  <a href="../articles/netReg_commandline.html">C++</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://www.simon-dirmeier.net">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/dirmeier/netReg">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="http://twitter.com/simon_dirmeier/">
    <span class="fa fa fa fa-twitter-square fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>netReg commandline tutorial</h1>
            
          </div>

    
    
<div class="contents">
<p>This document explains how to install and use the <code>netReg</code> command line tool. If something is unclear/buggy/not working I am glad to take any questions.</p>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>You have two options for installing the command line tool. You can either</p>
<ul>
<li>use <a href="https://anaconda.org/bioconda/netreg">bioconda</a>,</li>
<li>or install the tool manually.</li>
</ul>
<div id="installation-with-bioconda" class="section level3">
<h3 class="hasAnchor">
<a href="#installation-with-bioconda" class="anchor"></a>Installation with bioconda</h3>
<p>The best way to install <code>netReg</code> is using <em>conda</em>. For that you should download <a href="https://www.continuum.io/downloads">Anaconda</a> and create a <a href="https://conda.io/docs/using/envs.html">virtual environment</a>. Then install the tool using:</p>
<div class="sourceCode"><pre class="sourceCode sh"><code class="sourceCode bash">$ <span class="ex">conda</span> install -c bioconda netreg
  
$ <span class="ex">netReg</span> -h</code></pre></div>
<p>This would install the package from the <em>bioconda</em> channel (a highly recommented platform) and then test if the tool is installed.</p>
</div>
<div id="manual-installation" class="section level3">
<h3 class="hasAnchor">
<a href="#manual-installation" class="anchor"></a>Manual installation</h3>
<p>Alternatively you can of course install <code>netReg</code> manually. The <strong>command line tool has the following dependencies</strong>:</p>
<ul>
<li>
<code>CMake &gt;= 3.6</code>,</li>
<li>
<code>Boost &gt;= 1.6.x</code>,</li>
<li>
<code>Armadillo &gt;= 7.800.3</code>,</li>
<li>
<code>OpenBLAS/BLAS</code> and <code>Lapack</code> (older versions should work),</li>
<li>
<em>optional</em>: <code>OpenMP</code> (older versions should work).</li>
</ul>
<p>Then download the tarball of the latest <a href="https://github.com/dirmeier/netReg/releases/">release</a>. We use <code>CMake</code> for building the package. You build the package like this:</p>
<div class="sourceCode"><pre class="sourceCode sh"><code class="sourceCode bash">$ <span class="fu">mkdir</span> build <span class="kw">&amp;&amp;</span> <span class="bu">cd</span> build
$ <span class="fu">cmake</span> .. <span class="kw">&amp;&amp;</span> <span class="fu">make</span>
  
$ <span class="ex">./netReg</span> -h</code></pre></div>
<p>This builds the package in folder <code>build</code> and then tests if the package is installed. In case you want to install it properly subsequently use:</p>
<div class="sourceCode"><pre class="sourceCode sh"><code class="sourceCode bash">$ <span class="kw">(</span><span class="fu">sudo</span><span class="kw">)</span> <span class="fu">make</span> install --prefix=/some/path</code></pre></div>
</div>
</div>
<div id="usage" class="section level2">
<h2 class="hasAnchor">
<a href="#usage" class="anchor"></a>Usage</h2>
<p>This section explains how to fit a linear model and do parameter estimation. Using the command line tool is rather straightforward. First let’s call <code>netReg</code> without parameters to see how it is used:</p>
<div class="sourceCode"><pre class="sourceCode sh"><code class="sourceCode bash">$ <span class="ex">netReg</span> -h
  
  <span class="ex">netReg</span> - a network-regularized generalized regression model

  <span class="ex">Arguments</span>:
    <span class="ex">-h</span> [ --help ]                        Print this help.
    <span class="ex">-d</span> [ --design ] arg                  Filename of the design matrix X.
    <span class="ex">-u</span> [ --gx ] arg                      Filename of the affinity matrix GX
                                         <span class="kw">for</span> <span class="ex">X.</span>
    <span class="ex">-r</span> [ --reponse ] arg                 Filename of the reponse matrix Y.
    <span class="ex">-v</span> [ --gy ] arg                      Filename of the affinity matrix GY 
                                         <span class="kw">for</span> <span class="ex">Y.</span>
    <span class="ex">-l</span> [ --lambda ] arg (=1)             <span class="ex">LASSO</span> penalization parameter.
    <span class="ex">-s</span> [ --psi ] arg (=0)                <span class="ex">Penalization</span> parameter for affinity
                                         <span class="ex">matrix</span> GX.
    <span class="ex">-x</span> [ --xi ] arg (=0)                 <span class="ex">Penalization</span> parameter for affinity
                                         <span class="ex">matrix</span> GY.
    <span class="ex">-m</span> [ --maxit ] arg (=100000)         <span class="ex">Maximum</span> number of iterations of
                                         <span class="ex">coordinate</span> descent. You should 
                                         <span class="ex">choose</span> a sufficiently large number.
    <span class="ex">-t</span> [ --threshold ] arg (=0.0000001)  <span class="ex">Convergence</span> threshold for coordinate
                                         <span class="ex">descent.</span> Anything below 0.0001 
                                         <span class="ex">should</span> suffice.
    <span class="ex">-o</span> [ --outfile ] arg                 Filename of the output file.
    
    <span class="ex">Model</span> selection:
    <span class="ex">--modelselection</span>                     Use modelselection, i.e. estimation of
                                         <span class="ex">optimal</span> shrinkage parameters using
                                         <span class="ex">crossvalition</span>, before doing the
                                         <span class="ex">estimation</span> of coefficients.
    <span class="ex">-n</span> [ --nfolds ] arg (=10)            <span class="ex">The</span> number of cross-validation 
                                         <span class="ex">folds.</span> This can be maximal the number
                                         <span class="ex">of</span> rows of X/Y and minimal 3.
    <span class="ex">-e</span> [ --epsilon ] arg (=0.001)        <span class="ex">Convergence</span> threshold for the BOBYQA
                                         <span class="ex">algorithm</span>, i.e. the stop criterion for
                                         <span class="ex">the</span> model selection.
    <span class="ex">-b</span> [ --bobit ] arg (=1000)           <span class="ex">Maximal</span> number of iterations for the
                                         <span class="ex">BOBYQA</span> algorithm.</code></pre></div>
<p>This gives a list of required arguments which I think are rather self explanatory, but we will look at some examples later.</p>
<p>We have provided some example data files in the GitHub repository <a href="https://github.com/dirmeier/netReg/tree/master/data">here</a> to test the package. Files <code>X.tsv</code> and <code>Y.tsv</code> are taken from and adopted from a yeast eQTL study by <span class="citation">(Brem et al. 2005)</span>, <span class="citation">(Storey, Akey, and Kruglyak 2005)</span>, and <span class="citation">(Cheng et al. 2014)</span>. <code>GY.tsv</code> is taken from <a href="https://thebiogrid.org/downloads/archives/Release%20Archive/BIOGRID-3.4.150/BIOGRID-ORGANISM-3.4.150.tab.zip">BioGRID</a>. <code>X.tsv</code> is a <span class="math inline">\((n \times p)\)</span> matrix of genetic markers where <span class="math inline">\(n\)</span> is the number of samples (112) and <span class="math inline">\(p\)</span> is the number of markers. <code>Y.tsv</code> is a <span class="math inline">\((n \times q)\)</span> matrix of expression values for <span class="math inline">\(q\)</span> yeast genes. <span class="math inline">\(n\)</span> is again the numer of samples (112). <code>GY.tsv</code> is a <span class="math inline">\((q \times q)\)</span> adjacency matrix representing protein-protein interactions for the <span class="math inline">\(q\)</span> response variables.</p>
<div id="model-fitting" class="section level3">
<h3 class="hasAnchor">
<a href="#model-fitting" class="anchor"></a>Model fitting</h3>
<p>An exemplary call could look like this:</p>
<div class="sourceCode"><pre class="sourceCode sh"><code class="sourceCode bash">$ <span class="ex">netReg</span> -d X.tsv -r Y.tsv -v GY.tsv 
         <span class="ex">-l</span> 10 -x 1 --maxit 1000 --threshold 0.0001 
         <span class="ex">-o</span> outfile.tsv</code></pre></div>
<p>In this example we use a design matrix given in <code>X.tsv</code> and response matrix <code>Y.tsv</code>. For the response variables we use an affinity matrix that represents <em>biological relationships</em> with <code>GY.tsv</code>. We promote sparsity by setting <span class="math inline">\(\lambda = 10\)</span> and put a week penalty on similar coefficients by setting <span class="math inline">\(\xi = 1\)</span>. We stop the algorithm after maximal <span class="math inline">\(1000\)</span> iterations or if a convergence threshold of <span class="math inline">\(0.0001\)</span> is reached. The coefficients are written to <code>outfile.tsv</code>.</p>
</div>
<div id="model-selection" class="section level3">
<h3 class="hasAnchor">
<a href="#model-selection" class="anchor"></a>Model selection</h3>
<p>Usually we don’t know these parameters, but need to estimate them such that the correct parameterization gives us the <em>best</em> model. In that case we would call</p>
<div class="sourceCode"><pre class="sourceCode sh"><code class="sourceCode bash">$ <span class="ex">netReg</span> --modelselection 
         <span class="ex">-d</span> X.tsv -r Y.tsv -v GY.tsv 
         <span class="ex">-b</span> 100 
         <span class="ex">-o</span> outfile.tsv</code></pre></div>
<p>This estimates <span class="math inline">\(\lambda\)</span> and <span class="math inline">\(\xi\)</span> and writes them to <code>outfile.tsv</code>. Since we did not provide an affinity matrix for the design matrix we do not estimate <span class="math inline">\(\psi\)</span>. The optimization uses the <code>BOBYQA</code> algorithm by Powell <span class="citation">(Powell 2009)</span>. Since for such a complex likelihood as we use it this takes time, we only iterate <code>BOBYQA</code> <span class="math inline">\(100\)</span> times here. Other than that we use default parameters here.</p>
</div>
</div>
<div id="references" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<div id="refs" class="references">
<div id="ref-brem2005genetic">
<p>Brem, Rachel B, John D Storey, Jacqueline Whittle, and Leonid Kruglyak. 2005. “Genetic Interactions Between Polymorphisms That Affect Gene Expression in Yeast.” <em>Nature</em> 436 (7051). NIH Public Access: 701.</p>
</div>
<div id="ref-cheng2014graph">
<p>Cheng, Wei, Xiang Zhang, Zhishan Guo, Yu Shi, and Wei Wang. 2014. “Graph-Regularized Dual Lasso for Robust eQTL Mapping.” <em>Bioinformatics</em> 30 (12). Oxford Univ Press: i139–i148.</p>
</div>
<div id="ref-powell2009bobyqa">
<p>Powell, Michael JD. 2009. “The Bobyqa Algorithm for Bound Constrained Optimization Without Derivatives.” <em>Cambridge NA Report Na2009/06, University of Cambridge, Cambridge</em>. Citeseer.</p>
</div>
<div id="ref-storey2005multiple">
<p>Storey, John D, Joshua M Akey, and Leonid Kruglyak. 2005. “Multiple Locus Linkage Analysis of Genomewide Expression in Yeast.” <em>PLoS Biology</em> 3 (8). Public Library of Science: e267.</p>
</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#installation">Installation</a></li>
      <li><a href="#usage">Usage</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Simon Dirmeier.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
